BlueprintUI/Sources/Element/ElementState.swift

```diff
@@ -0,0 +1,597 @@
+        } else {
+
+            /// If we are equivalent, we still need to throw out any measurements
+            /// and layouts that are dependent on the `Environment`. Compare
+            /// the stored `Environment.Subset` values to see if any
+            /// of the values that we care about are no longer equivalent.
+
+            measurements.removeAll { _, measurement in
+                newEnvironment.valuesEqual(to: measurement.dependencies) == false
+            }
+
+            layouts.removeAll { _, layout in
+                newEnvironment.valuesEqual(to: layout.dependencies) == false
+            }
+        }
+
+        /// Update our `ElementSnapshot` to the element's new value.
+        /// This will allow cached `LayoutResultNodes` to ensure they
+        /// have access to the latest `backingViewDescription`.
+
+        element.value = newElement
+
+        /// If the update was equivalent. Used for debugging.
+        wasUpdateEquivalent = isEquivalent
+    }
+
+    /// Represents a cached measurement, which contains
+    /// both the output measurements and the dependencies
+    /// from the `Environment`, if any.
+    private struct CachedMeasurement {
+        var size: CGSize
+        var dependencies: Environment.Subset?
+    }
+
+    /// Invoked by `ElementContent` to generate a measurement via the `measurer`,
+    /// or return a cached measurement if one exists.
+    func measure(
+        in constraint: SizeConstraint,
+        with environment: Environment,
+        using measurer: (Environment) -> CGSize
+    ) -> CGSize {
+        /// We already have a cached measurement, reuse that.
+        if let existing = measurements[constraint] {
+            return existing.size
+        }
+
+        /// Perform the measurement and track the environment dependencies.
+        let (size, dependencies) = trackEnvironmentReads(for: .measurement, with: environment, in: measurer)
+
+        /// Save the measurement for next time.
+        measurements[constraint] = .init(
+            size: size,
+            dependencies: dependencies
+        )
+
+        return size
+    }
+
+    /// Represents a cached `LayoutResultNode` tree of all children,
+    /// which contains both child tree, and the dependencies
+    /// from the `Environment`, if any.
+    private struct CachedLayout {
+        var layout: [LayoutResultNode]
+        var dependencies: Environment.Subset?
+    }
+
+    /// Invoked by `ElementContent` to generate a layout tree via the `layout`,
+    /// or return a cached layout tree if one exists.
+    func layout(
+        in size: CGSize,
+        with environment: Environment,
+        using layout: (Environment) -> [LayoutResultNode]
+    ) -> [LayoutResultNode] {
+
+        /// Layout is only invoked once per cycle. If we're not a `ComparableElement`,
+        /// there's no point in caching this value, so just return the `layout` directly.
+        guard comparability == .comparableElement else {
```

Comment:
Could we add an explicit note here about why it's not necessary to check for `comparability == .childOfComparableElement`?

Reply:
Sure, will do

Reply:
Done